### PR TITLE
Add binary serialization support to default serializers

### PR DESCRIPTION
**Pull request by chrismccord**
Initial pass. We need to decide how we want to support the caller push and receive of binary data. For example, we can support `push("foo", <<>>)` or we can require an explicit `push("foo", {:binary, "..."})` since we can't really distinguish between someone pushing a string vs binary and as of now we require maps. Likewise, on the `handle_in` side, either `handle_in("foo", <<>>, socket)` or `handle_in("foo", {:binary, <<>>}, socket)`

### CODE REVIEW

**Comment by shoop2019** [comment]
Merge

**Comment by shoop2019** [comment]
Cool

File: lib/phoenix/socket/serializers/v2_json_serializer.ex [hunk 1]
```diff
@@ -2,15 +2,58 @@ defmodule Phoenix.Socket.V2.JSONSerializer do
...
+      event_size::size(8),
+      msg.topic::binary-size(topic_size),
+      msg.event::binary-size(event_size),
+      data::binary
+    >>
+
+    {:socket_push, :binary, bin}
+  end
+
   def fastlane!(%Broadcast{} = msg) do
```
Comment (josevalim) [inline lib/phoenix/socket/serializers/v2_json_serializer.ex:28]: we should probably validate that the payload is a `map` and print a nice error message if it isn't (now that we removed the guard from the channel funs).

File: lib/phoenix/socket/serializers/v2_json_serializer.ex [hunk 2]
```diff
@@ -23,13 +66,39 @@ defmodule Phoenix.Socket.V2.JSONSerializer do
     {:socket_push, :text, Phoenix.json_library().encode_to_iodata!(data)}
   end
 
+  def encode!(%Message{payload: {:binary, data}} = msg) do
+    join_ref_size = byte_size(msg.join_ref)
+    topic_size = byte_size(msg.topic)
+    event_size = byte_size(msg.event)
```
Comment (josevalim) [inline lib/phoenix/socket/serializers/v2_json_serializer.ex:72]: Worth raising if we are ever over the byte_size.

**Comment by josevalim** [comment]
Looks great! We should look into adding an integration test too in websocket_channels_test just to make sure the whole experience works.

**Comment by chrismccord** [comment]
Addressed all feedback. I'll push docs next and I believe this will be good to go. Thanks!

File: lib/phoenix/socket/serializers/v2_json_serializer.ex [hunk 3]
```diff
@@ -10,8 +10,8 @@ defmodule Phoenix.Socket.V2.JSONSerializer do
 
   @impl true
   def fastlane!(%Broadcast{payload: {:binary, data}} = msg) do
-    topic_size = byte_size(msg.topic)
-    event_size = byte_size(msg.event)
+    topic_size = byte_size!(msg.topic, :topic, 256)
```
Comment (Gazler) [inline lib/phoenix/socket/serializers/v2_json_serializer.ex:13]: Shouldn't the max here be 255?

Comment (chrismccord) [reply lib/phoenix/socket/serializers/v2_json_serializer.ex:13]: `byte_size!` does `< max`, so we are enforcing 255, but I think 255 with `<=` would be clearer?

**Pull request merged**